main.c

```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
     }
 }
 
-struct bzimage_params {
-    size_t bzimage_addr;
+struct boot_params {
+    size_t kernel_addr;
     size_t initrd_addr;
     size_t initrd_size_addr;
     size_t cmdline_addr;
 };
 
-static struct bzimage_params initialize_from_fw_cfg()
+enum boot_protocol {
+    UNKNOWN,
+    LINUX,
+    ELF,
+};
+
+static struct boot_params get_boot_params_from_fw_cfg();
+static enum boot_protocol get_boot_protocol(struct boot_params);
+void linux_boot(struct boot_params);
+void elf_boot(struct boot_params);
+
+int main(void)
+{
+    struct boot_params params = get_boot_params_from_fw_cfg();
+    enum boot_protocol boot_protocol = get_boot_protocol(params);
+
+    switch (boot_protocol) {
+    case LINUX:
+        linux_boot(params);
+        break;
+    case ELF:
+        elf_boot(params);
+        break;
+    default:
+        die_on(true, "Unknown boot method\n");
+    }
+}
+
+// Identify the type of the kernel.
+static enum boot_protocol get_boot_protocol(struct boot_params params)
+{
+    size_t kernel_addr = params.kernel_addr;
+
+    die_on(kernel_addr == 0, "Did not find the address of kernel.\n");
+
+    // Linux kernel supporting the "new" boot protocol have a magic number at a specific offset.
+    // See https://www.kernel.org/doc/Documentation/x86/boot.txt.
+    if (*(uint32_t *)(kernel_addr + 0x202) == 0x53726448) {
+        return LINUX;
+    }
+
+    // ELF binaries begin with a magic number.
+    if (*(uint32_t *)kernel_addr == 0x464c457f) {
```

Comment:
Should this magic number live in `elf_boot.h`?

Reply:
You are right - I didn't think of it.